lib/datadog/core/normalizer.rb

```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+module Datadog
+  module Core
+    module Normalizer
+      module_function
+
+      INVALID_TAG_CHARACTERS = %r{[^\p{L}0-9_\-:./]}.freeze
+      LEADING_INVALID_CHARS = %r{\A[^\p{L}:]+}.freeze
+      TRAILING_UNDERSCORES = %r{_+\z}.freeze
+      MAX_CHARACTER_LENGTH = (0...200).freeze
+
+      # Based on https://github.com/DataDog/datadog-agent/blob/45799c842bbd216bcda208737f9f11cade6fdd95/pkg/trace/traceutil/normalize.go#L131
+      # Specifically:
+      # - Must be valid UTF-8
+      # - Invalid characters are replaced with an underscore
+      # - Leading non-letter characters are removed but colons are kept
+      # - Trailing non-letter characters are removed
+      # - Trailing underscores are removed
+      # - Consecutive underscores are merged into a single underscore
+      # - Maximum length is 200 characters
+      def self.normalize(original_value)
+        normalized_value = original_value.to_s.encode('UTF-8', invalid: :replace, undef: :replace).strip
```

Comment:
I guess we could use `strip!` here to save allocation as encode will return you a dup of the string

```c
// ... Ruby source ...
return encoded_dup(newstr, str, encidx);
```

Reply:
Addressed in https://github.com/DataDog/dd-trace-rb/pull/5033/commits/41bc6c0e966edee743063bc3de99c8bd9c6ddcc9!